##Методические указания

In [74]:
import warnings
warnings.filterwarnings("ignore")

In [75]:
import pandas as pd


In [76]:
data = pd.read_csv('heart.csv')


In [77]:
y = data["output"]
x = data.drop("output", axis=1)

from sklearn.linear_model import LogisticRegression

logistic = LogisticRegression().fit(x, y)
logistic.score(x, y)

0.8547854785478548

In [78]:
x_train, y_train = x[:200], y[:200]

In [79]:
x_train.shape, y_train.shape

((200, 13), (200,))

In [80]:
x_test, y_test = x[200:], y[200:]
x_test.shape, y_test.shape

((103, 13), (103,))

In [81]:
logistic_test = LogisticRegression().fit(x_train, y_train)
logistic_test.score(x_train, y_train), logistic_test.score(x_test, y_test)

(0.9, 0.5436893203883495)

In [82]:
N = int(x.shape[0] * 0.8)

x_train, y_train, x_test, y_test = x[:N], y[:N], x[N:], y[N:]
x_train.shape, y_train.shape, x_test.shape, y_test.shape

((242, 13), (242,), (61, 13), (61,))

In [83]:
logistic_test = LogisticRegression().fit(x_train, y_train)
logistic_test.score(x_train, y_train), logistic_test.score(x_test, y_test)

(0.8884297520661157, 0.6229508196721312)

In [84]:
data.tail()

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
298   57    0   0     140   241    0        1       123     1      0.2    1   
299   45    1   3     110   264    0        1       132     0      1.2    1   
300   68    1   0     144   193    1        1       141     0      3.4    1   
301   57    1   0     130   131    0        1       115     1      1.2    1   
302   57    0   1     130   236    0        0       174     0      0.0    1   

     caa  thall  output  
298    0      3       0  
299    0      3       0  
300    2      3       0  
301    1      3       0  
302    1      2       0

In [85]:
import numpy as np

In [86]:
mask = np.array([True] * N + [False] * (y.shape[0] - N))

In [87]:
from numpy.random import shuffle

shuffle(mask)
mask

array([ True,  True,  True,  True, False,  True, False,  True,  True,
        True,  True,  True, False,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True, False,  True, False,  True,
        True,  True, False,  True, False,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
       False,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True, False,  True, False,  True, False, False,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
       False,  True,  True,  True,  True, False,  True,  True,  True,
        True,  True, False,  True,  True,  True,  True,  True, False,
        True,  True, False,  True,  True,  True,  True,  True,  True,
       False,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True, False,  True,  True,  True,  True,  True,
        True, False,  True,  True, False, False,  True, False,  True,
        True,  True,

In [88]:
x_train = x[mask]
x_train.shape

(242, 13)

In [89]:
x_train, y_train, x_test, y_test = x[mask], y[mask], x[~mask], y[~mask]
x_train.shape, y_train.shape, x_test.shape, y_test.shape

((242, 13), (242,), (61, 13), (61,))

In [90]:
logistic_test = LogisticRegression().fit(x_train, y_train)
logistic_test.score(x_train, y_train), logistic_test.score(x_test, y_test)

(0.8636363636363636, 0.8360655737704918)

In [91]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x, y, train_size=0.8)
x_train.shape, y_train.shape, x_test.shape, y_test.shape

((242, 13), (242,), (61, 13), (61,))

In [92]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix, classification_report

In [93]:
y_test_pred = logistic_test.predict(x_test)
y_train_pred = logistic_test.predict(x_train)

In [94]:
confusion_matrix(y_train, y_train_pred)

array([[ 85,  26],
       [  7, 124]])

In [95]:
confusion_matrix(y_test, y_test_pred)

array([[20,  7],
       [ 3, 31]])

In [96]:
print(classification_report(y_test, y_test_pred))

              precision    recall  f1-score   support

           0       0.87      0.74      0.80        27
           1       0.82      0.91      0.86        34

    accuracy                           0.84        61
   macro avg       0.84      0.83      0.83        61
weighted avg       0.84      0.84      0.83        61



In [97]:
precision_score(y_test, y_test_pred)

0.8157894736842105

In [98]:
metrics = pd.DataFrame({
    "Train": [
        accuracy_score(y_train, y_train_pred),
        precision_score(y_train, y_train_pred),
        recall_score(y_train, y_train_pred),
        f1_score(y_train, y_train_pred),
    ],
    "Test": [
        accuracy_score(y_test, y_test_pred),
        precision_score(y_test, y_test_pred),
        recall_score(y_test, y_test_pred),
        f1_score(y_test, y_test_pred),
    ],
}, index = ["Accuracy", "Precision", "Recall", "F1"])

metrics

Train      Test
Accuracy   0.863636  0.836066
Precision  0.826667  0.815789
Recall     0.946565  0.911765
F1         0.882562  0.861111

##Задания для самостоятельного выполнения

#### 1. Повторите анализ для других видов моделей. Используйте 5-10 разных классов моделей. Подсчитывайте только метрики на тестовой выборке.

In [99]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.neural_network import MLPClassifier


X = data.drop('output', axis=1)
y = data['output']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=19)

classifiers = {
    'Logistic Regression': LogisticRegression(),
    'Decision Tree': DecisionTreeClassifier(),
    'Random Forest': RandomForestClassifier(),
    'SVM': SVC(),
    'K Nearest Neighbors': KNeighborsClassifier(),
    'Naive Bayes': GaussianNB(),
    'Neural Network': MLPClassifier()
}

# Обучение и оценка эффективности каждой модели
results = {}
for name, clf in classifiers.items():
    clf.fit(X_train, y_train)
    y_pred = clf.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred)
    recall = recall_score(y_test, y_pred)
    f1 = f1_score(y_test, y_pred)
    results[name] = {'Accuracy': accuracy, 'Precision': precision, 'Recall': recall, 'F1-score': f1}

results_df = pd.DataFrame(results)
print(results_df)


           Logistic Regression  Decision Tree  Random Forest       SVM  \
Accuracy              0.786885       0.754098       0.770492  0.590164   
Precision             0.771429       0.814815       0.800000  0.581395   
Recall                0.843750       0.687500       0.750000  0.781250   
F1-score              0.805970       0.745763       0.774194  0.666667   

           K Nearest Neighbors  Naive Bayes  Neural Network  
Accuracy              0.622951     0.754098        0.786885  
Precision             0.628571     0.793103        0.743590  
Recall                0.687500     0.718750        0.906250  
F1-score              0.656716     0.754098        0.816901  


Модели Decision Tree, Random Forest и Neural Network показывают сравнимые результаты по точности (Accuracy), прецизии (Precision) и F1-мере (F1-score). Они имеют промежуточные значения между другими моделями.

На основе метрик качества, модели Logistic Regression, Decision Tree, Random Forest и Neural Network можно рассматривать как наиболее перспективные для данной задачи классификации

#### 2. Повторите анализ для другого датасета по вашему выбору. Используйте несколько моделей для сравнения. Используйте датасет для множественной классификации.

In [72]:
from sklearn.datasets import load_wine

wine = load_wine()
X = wine.data
y = wine.target

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=19)

classifiers = {
    'Logistic Regression': LogisticRegression(),
    'Decision Tree': DecisionTreeClassifier(),
    'Random Forest': RandomForestClassifier(),
    'SVM': SVC(),
    'K Nearest Neighbors': KNeighborsClassifier(),
    'Naive Bayes': GaussianNB(),
    'Neural Network': MLPClassifier()
}

results = {}
for name, clf in classifiers.items():
    clf.fit(X_train, y_train)
    y_pred = clf.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred, average='weighted')
    recall = recall_score(y_test, y_pred, average='weighted')
    f1 = f1_score(y_test, y_pred, average='weighted')
    results[name] = {'Accuracy': accuracy, 'Precision': precision, 'Recall': recall, 'F1-score': f1}

results_df = pd.DataFrame(results)
print(results_df)


           Logistic Regression  Decision Tree  Random Forest       SVM  \
Accuracy              0.916667       0.833333       0.944444  0.694444   
Precision             0.920486       0.885462       0.954545  0.679398   
Recall                0.916667       0.833333       0.944444  0.694444   
F1-score              0.917597       0.831234       0.945486  0.684195   

           K Nearest Neighbors  Naive Bayes  Neural Network  
Accuracy              0.666667     0.944444        0.916667  
Precision             0.653722     0.954545        0.920486  
Recall                0.666667     0.944444        0.916667  
F1-score              0.654434     0.945486        0.917597  


Снова модель Random Forest демонстрирует высокие показатели точности (accuracy) и precision (точность), а также F1-score. Ее accuracy составляет 94.4%, что делает ее лучшим выбором среди рассмотренных моделей.

Модель Logistic Regression также показывает высокие результаты, с точностью, recall и F1-score близкими к 92%. Она является вторым лучшим выбором после модели Random Forest.

Итак, на основании метрик качества, модель Random Forest снова является наилучшим выбором для данной задачи классификации

####3. Повторите анализ для датасета, предназначенного для решения задачи регрессии. Используйте все метрики качества регрессии, изученные на лекции. Постройте 5 - 10 разных моделей регрессии.

In [73]:
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.svm import SVR
from sklearn.neighbors import KNeighborsRegressor
from sklearn.linear_model import Lasso, Ridge
from sklearn.preprocessing import StandardScaler
from sklearn.datasets import fetch_california_housing

california_housing = fetch_california_housing()
X = california_housing.data
y = california_housing.target

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=19)

# Стандартизация признаков
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

regressors = {
    'Linear Regression': LinearRegression(),
    'Decision Tree': DecisionTreeRegressor(),
    'Random Forest': RandomForestRegressor(),
    'SVR': SVR(),
    'K Neighbors': KNeighborsRegressor(),
    'Lasso': Lasso(),
    'Ridge': Ridge()
}

# Обучение и оценка эффективности каждой модели
results = {}
for name, reg in regressors.items():
    reg.fit(X_train_scaled, y_train)
    y_pred = reg.predict(X_test_scaled)
    mse = mean_squared_error(y_test, y_pred)
    mae = mean_absolute_error(y_test, y_pred)
    r2 = r2_score(y_test, y_pred)
    results[name] = {'Mean Squared Error': mse, 'Mean Absolute Error': mae, 'R^2 Score': r2}

results_df = pd.DataFrame(results)
print(results_df)


                     Linear Regression  Decision Tree  Random Forest  \
Mean Squared Error            0.530318       0.522361       0.255323   
Mean Absolute Error           0.532540       0.463977       0.335722   
R^2 Score                     0.600171       0.606171       0.807502   

                          SVR  K Neighbors     Lasso     Ridge  
Mean Squared Error   0.323228     0.397094  1.326418  0.530294  
Mean Absolute Error  0.384593     0.433275  0.911943  0.532533  
R^2 Score            0.756305     0.700614 -0.000041  0.600190  


Модель Random Forest демонстрирует самые низкие значения среднеквадратичной ошибки (Mean Squared Error) и средней абсолютной ошибки (Mean Absolute Error), а также наивысший коэффициент детерминации (R^2 Score). Это говорит о том, что модель Random Forest лучше всего приближает зависимость между признаками и целевой переменной в данном наборе данных.

Модель SVR (Support Vector Regression) также показывает неплохие результаты, хотя немного уступает по всем метрикам модели Random Forest.

Итак, на основе метрик качества, модель Random Forest является лучшим выбором для данной задачи регрессии